File: app/src/main/java/com/example/administrator/okhttputil/net/response/CommonJsonCallback.java
package com.example.administrator.okhttputil.net.response;

import android.os.Handler;
import android.os.Looper;

import com.example.administrator.okhttputil.net.exception.OkHttpException;
import com.example.administrator.okhttputil.net.listener.DisposeDataHandle;
import com.example.administrator.okhttputil.net.listener.DisposeDataListener;
import com.google.gson.Gson;

import org.json.JSONObject;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Response;

/**
 * Create by SunnyDay on 2019/03/05
 */
public class CommonJsonCallback implements Callback {
    protected final String RESULT_CODE = "ecode";
    protected final int RESULT_CODE_VALUE = 0;
    protected final String ERROR_MSG = "emsg";
    protected final String EMPTY_MSG = "";

    /**
     * 自定义类型异常
     */
    protected final int NETWORK_ERROR = -1;
    protected final int JSON_ERROR = -2;
    protected final int OTHER_ERROR = -3;


    private Handler mDeliveryHandler; //进行消息转发
    private DisposeDataListener mlistener;
    private Class<?> mClass;

    public CommonJsonCallback(DisposeDataHandle handle) {
        this.mlistener = handle.listener;
        this.mClass = handle.mClass;
        this.mDeliveryHandler = new Handler(Looper.getMainLooper());
    }

    /**
     * 请求失败处理
     */
    @Override
    public void onFailure(Call call, final IOException e) {
        mDeliveryHandler.post(new Runnable() {
            @Override
            public void run() {
                mlistener.onFailure(new OkHttpException(NETWORK_ERROR, e));
            }
        });
    }

    /**
     * 真正的响应处理
     */
    @Override
    public void onResponse(Call call, Response response) throws IOException {
        final String result = response.body().string();// json 类型 为String就行
        mDeliveryHandler.post(new Runnable() {
            @Override
            public void run() {
                handleResponse(result);
            }
        });
    }

    private void handleResponse(Object responseObj) {
        if (responseObj == null && responseObj.toString().trim().equals("")) {
            mlistener.onFailure(new OkHttpException(NETWORK_ERROR, EMPTY_MSG));
            return;
        }
        try {
            JSONObject result = new JSONObject(responseObj.toString());
            if (result.has(RESULT_CODE)) {
                if (result.getInt(RESULT_CODE) == RESULT_CODE_VALUE) {
                    if (mClass == null) {
                        mlistener.onSuccess(responseObj);
                    } else {
                        // 转化为实体
                        Gson gson = new Gson();
                        Object obj = gson.fromJson(responseObj.toString(), mClass);
                        if (obj != null) {
                            mlistener.onSuccess(obj);
                        } else {
                            // 不是合法的json
                            mlistener.onFailure(new OkHttpException(JSON_ERROR, EMPTY_MSG));
                        }
                    }
                }
            }else{
                 mlistener.onFailure(new OkHttpException(OTHER_ERROR,result.get(RESULT_CODE)));
            }
        } catch (Exception e) {
           mlistener.onFailure(new OkHttpException(OTHER_ERROR,e.getMessage()));
        }
    }
}

File: app/src/main/java/com/example/administrator/okhttputil/ktversion/response/CommonJsonCallback.kt
package com.example.selfupdate.net.response

import android.os.Handler
import android.os.Looper
import com.example.selfupdate.net.exception.CommonException
import com.example.selfupdate.net.listener.DisposeDataHandler
import com.example.selfupdate.net.listener.DisposeDataListener
import com.example.selfupdate.beans.AppUpdateInfo
import com.google.gson.Gson
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response
import java.io.IOException
import java.lang.Exception

/**
 * Created by zb on 2021/3/31 16:00
 *
 * CommonJson callBack, Convert stream data to json String !
 */

class CommonJsonCallback(
    disposeDataHandler: DisposeDataHandler
) : Callback {

    companion object{
        private const val NETWORK_ERROR: Int = -1
        private const val JSON_ERROR: Int = -2
        private const val OTHER_ERROR: Int = -3
    }

    private var listener: DisposeDataListener = disposeDataHandler.listener
    private var mHandler: Handler = Handler(Looper.getMainLooper())
    private val mClass: Class<*>? = disposeDataHandler.clazz


    override fun onFailure(call: Call, e: IOException?) {
        mHandler.post {
            listener.onFailure(CommonException(NETWORK_ERROR, "net work error!"))
        }
    }

    override fun onResponse(call: Call, response: Response?) {
        val json = response!!.body().string()
        mHandler.post {
            handleStringJson(json)
        }
    }


    private fun handleStringJson(jsonStr: String) {

        if (mClass == null) {
            listener.onSuccess(jsonStr)
        } else {

            try {
                val gson = Gson()
                val obj = gson.fromJson(jsonStr, mClass)
                if (obj != null) {
                    listener.onSuccess(obj)
                } else {
                    listener.onFailure(CommonException(JSON_ERROR, "parse json error!"))
                }
            } catch (e: Exception) {
                e.printStackTrace()
                listener.onFailure(CommonException(JSON_ERROR, "parse json error!"))
            }

        }
    }
}